src/deim.jl

```diff
@@ -83,7 +83,17 @@ function deim(sys::ODESystem, pod_basis::AbstractMatrix;
 
     pod_dict = Dict(eq.lhs => eq.rhs for eq in pod_eqs) # original vars to reduced vars
 
-    U = @view deim_basis[:, 1:deim_dim] # DEIM projection basis
+    # generate an in-place function from the symbolic expression of the nonlinear functions
+    F_expr = build_function(F, dvs; expression = Val{false})[2]
+    F_func! = eval(F_expr)
+    nonlinear_snapshot = similar(snapshot) # snapshot matrix of nonlinear terms
+    for i in 1:size(snapshot, 2) # iterate through time instances
+        F_func!(view(nonlinear_snapshot, :, i), view(snapshot, :, i))
+    end
+
+    deim_reducer = POD(nonlinear_snapshot, deim_dim)
+    reduce!(deim_reducer, TSVD())
+    U = deim_reducer.rbasis # DEIM projection basis
 
     indices = deim_interpolation_indices(U) # DEIM interpolation indices
     # the DEIM projector (not DEIM basis) satisfies
```

Comment:
The changes look good but would you mind also adding a "lower level" deim method with a signature like

function deim(variables, linear_part,  constant_part, nonlinear_part, projection_basis, interpolation_basis; kwargs...)
    ....
end
where we supply the all the pieces individually. That sets us up to combine this with other reduction techniques (for example, there are good reason for not wanting to project onto the POD but a different basis). I am not aware what other bases would be sensible to use for interpolation but it is at least conceivable. Also this way, we make what you have build accessible even if the automatic extraction of linear and nonlinear parts is too expensive.

There are several potential objects that this function could return. I think it is most natural to either return the reduced model equations or an ODESystem.

Reply:
I guess such a function should be named something like `nonlinear_projection_reduce`. 

Typically one doesn't immediately have the full set of ODEs available, and thus applies ModelingToolkit to, for example, 
- build hierarchical component-based models,
- use the method of lines to automatically discretize PDEs,
- call `ModelingToolkit.structural_simplify`, etc.

Therefore, I think we would better not expect users to supply nonlinear terms.

`Symbolics.semipolynomial_form` traverses all nodes in the expression tree and creates an object for each node. If the input expression has $n$ nodes, then `Symbolics.semipolynomial_form` takes $\Omega(n)$ time. `ModelOrderReduction.linear_terms` doesn't expand the input expression and only evaluates the operation and arguments of the root node of the expression tree, so takes $\mathcal O(1)$ constant time.

Reply:
Sorry for the delay. I just added this. in commit https://github.com/SciML/ModelOrderReduction.jl/pull/19/commits/68b521c99b38439aa032c63e0428993005ae7eab.